### PR TITLE
dumpling: fix s3 compress writer

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/30534

Problem Summary: fix the problem that dumpling can't dump with --compress and s3 output directory.

### Check List

Tests 

- [x] Unit test
- [x] Integration test

### Release note

```release-note
fix the problem that dumpling can't dump with --compress and s3 output directory.
```

### CODE REVIEW

File: dumpling/tests/naughty_strings/naughty_strings.escape-schema.sql [hunk 1]
```diff
@@ -1,4 +0,0 @@
-CREATE TABLE `escape` (
-  `a` text CHARACTER SET utf8mb4 COLLATE utf8mb4_bin
-) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
```
Comment (Ehco1996) [inline dumpling/tests/naughty_strings/naughty_strings.escape-schema.sql:3]: why delete this file?

File: br/pkg/storage/writer.go [hunk 2]
```diff
@@ -149,6 +155,10 @@ func (u *bufferedWriter) Write(ctx context.Context, p []byte) (int, error) {
 				return bytesWritten, errors.Trace(err)
 			}
 			p = p[w:]
+			// continue buf because compressed data size may be less than Cap - Len
+			if u.buf.Compressed() {
```
Comment (Ehco1996) [inline br/pkg/storage/writer.go:159]: can you explain more how this `continue` fix the error ? seems it just  save once `u.uploadChunk`

File: dumpling/tests/naughty_strings/naughty_strings.escape-schema.sql (hunk 1 above)
Comment (lichunzhu) [reply dumpling/tests/naughty_strings/naughty_strings.escape-schema.sql:3]: Because it's never used.

File: br/pkg/storage/writer.go (hunk 2 above)
Comment (lichunzhu) [reply br/pkg/storage/writer.go:159]: For normal files, writing n bytes means that `u.buf` will grow n bytes, so we can directly write chunkToFill bytes and u.buf can reach 5MB.
But for compressed files, writing n bytes means that `u.buf` will only grow m (m << n) bytes, now `u.buf` may still much less than 5MB, so we can upload here directly.

File: dumpling/tests/naughty_strings/naughty_strings.escape-schema.sql (hunk 1 above)
Comment (Ehco1996) [reply dumpling/tests/naughty_strings/naughty_strings.escape-schema.sql:3]: seems it still used in test https://github.com/pingcap/tidb/blob/321d307f3b23877fb2be0e7534a9ecd6a655be05/dumpling/tests/naughty_strings/run.sh#L23

**Comment by lance6716** [comment]
I prefer we implement a `LimitMinCapBuffer` and insert it between the level of `bytes.NewBuffer` and `CompressWriter`, so when `CompressWriter.Write` returns the written bytes less than expected or returns a specific error, we can rotate the upper writter.

maybe some logging libraries already have this technique?

File: dumpling/tests/naughty_strings/naughty_strings.escape-schema.sql (hunk 1 above)
Comment (lichunzhu) [reply dumpling/tests/naughty_strings/naughty_strings.escape-schema.sql:3]: This file's path is `dumpling/tests/naughty_strings/naughty_strings.escape-schema.sql` 🤔

Comment (Ehco1996) [reply dumpling/tests/naughty_strings/naughty_strings.escape-schema.sql:3]: ohohoh my bad... two file in different with same name

**Comment by lichunzhu** [comment]
> I prefer we implement a `LimitMinCapBuffer` and insert it between the level of `bytes.NewBuffer` and `CompressWriter`, so when `CompressWriter.Write` returns the written bytes less than expected or returns a specific error, we can rotate the upper writter.
> 
> maybe some logging libraries already have this technique?

I understand what you means. Can we merge this pr first as a bug fix and then file another issue as an improvement? Because for now the most important problem is that dumpling can work when upload compressed data to S3.

**Review by Ehco1996 — APPROVED** [review]

**Review by lance6716 — COMMENTED** [review]
where did we test that dumpling will generate files greater than 5MB?

**Comment by lichunzhu** [comment]
> where did we test that dumpling will generate files greater than 5MB?

We use this go file to import 50MB data into MySQL and dump them as one file in s3 test.
```
# load 50MB data into MySQL
(cd "$(dirname "$0")" && GO111MODULE=on go build -o out)
$DUMPLING_BASE_NAME/out -B $DB_NAME -T $TABLE_NAME -P 3306 -w 16
```

**Review by lance6716 — APPROVED** [review]

**Comment by lichunzhu** [comment]
related issue: https://github.com/pingcap/tidb/issues/30903

**Pull request merged**